sched/distributor/poolDistributor.go

```diff
@@ -1,180 +1,96 @@
+		Reserve:  make(chan cluster.Node),
+		Release:  make(chan cluster.Node),
+		updateCh: updateCh,
+		nodes:    nodes,
+		free:     free,
 	}
+	go dist.loop()
+	return dist
 }
 
-//
-// Creates a Pool Distributor, dynamic cluster &
-// all nodes in the cluster are free upon creation
-//
-// If the cluster has 0 nodes in it returns nil
-//
-func NewDynamicPoolDistributor(clusterState cm.DynamicClusterState) *PoolDistributor {
-
-	nodes := clusterState.InitialMembers
-
-	// allocate the initial cluster size for the free channel
-	// intializing the channel size to be 5 * the number of initial
-	// nodes.
-	//
-	// This is currently a bit of a hack for dynamic clusters
-	// adding nodes will just block, no updates will be lost.
-	// Doubling the cluster size is a rare event, and to fix, the scheduler
-	// just needs to be restarted for a larger channel to be allocated.
-	freeCh := make(chan cm.Node, 5*len(nodes))
-
-	// put all the nodes in the freeCh
-	for _, n := range nodes {
-		freeCh <- n
+func NewPoolDistributorFromCluster(cluster cluster.Cluster) (*PoolDistributor, error) {
+	sub, err := cluster.Subscribe()
+	if err != nil {
+		return nil, err
 	}
+	return NewPoolDistributor(sub.InitialMembers, sub.Updates), nil
+}
 
-	dist := &PoolDistributor{
-		freeCh:   freeCh,
-		toRemove: make(map[string]bool),
-	}
-
-	// spawn a go routine to deal with cluster updates
-	go updateCluster(dist, clusterState.Updates)
-
-	return dist
+func (d *PoolDistributor) Close() {
+	close(d.Release)
 }
 
-//
-// Watch for updates from the cluster and update internal list
-//
-func updateCluster(dist *PoolDistributor, updateCh <-chan cm.NodeUpdate) {
+// Serves pool distributor logic
+func (d *PoolDistributor) loop() {
 	for {
+		var reserveCh chan cluster.Node
+		var outgoing cluster.Node
+		if len(d.free) > 0 {
+			reserveCh, outgoing = d.Reserve, d.free[0]
+		}
 		select {
-
-		case nodeUpdate := <-updateCh:
-			switch nodeUpdate.UpdateType {
-			case cm.NodeAdded:
-				node := nodeUpdate.Node
-
-				// Check if this node is on the remove watch list,
-				// and remove it if it is.  This prevents flapping nodes
-				// from being removed after being re-added.
-				dist.mutex.RLock()
-				_, ok := dist.toRemove[node.Id()]
-				dist.mutex.RUnlock()
-
-				if ok {
-					dist.mutex.Lock()
-					// don't need to recheck condition because if the
-					// element has already been removed from the watchlist
-					// retrying to remove it is a safe operation, does nothing.
-					delete(dist.toRemove, node.Id())
-					dist.mutex.Unlock()
-				}
-
-				// add the node to the free queue
-				dist.freeCh <- node
-			case cm.NodeRemoved:
-				node := nodeUpdate.Node
-
-				// add the node to the Remove Watch List
-				dist.mutex.Lock()
-				dist.toRemove[node.Id()] = true
-				dist.mutex.Unlock()
+		case reserveCh <- outgoing:
+			d.free = d.free[1:]
+		case node, ok := <-d.Release:
+			if !ok {
+				return
+			}
+			if _, ok := d.nodes[node.Id()]; ok {
```

Comment:
Might as well do the !ok case, possibly panic depending on if we think this should happen or not. Actually, have we thought about how to make our code extra strict in dev/staging, if not all the time?